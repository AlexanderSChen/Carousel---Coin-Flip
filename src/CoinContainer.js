import React, { useState } from 'react';
import Coin from './Coin';
import './CoinContainer.css';
import { choice } from './helpers';

/** Coin container holds state about flip: heads/tails and current coin */

function CoinContainer(props) {
    const [coin, setCoin] = useState(null);
    const [headCount, setHeadCount] = useState(0);
    const [tailCount, setTailCount] = useState(0);
    const handleClick = () => {
        const newCoin = choice(props.coins);
        setCoin(newCoin);
        if (newCoin.side === "head") {
            setHeadCount(oldCount => oldCount + 1);
        } else {
            setTailCount(oldCount => oldCount + 1);
        }
    };

    const currCoin = coin ? (
        <Coin side={coin.side} imgSrc = {coin.imgSrc} />
    ) : null;

    return (
        <div className="CoinContainer">
            <h2>Flip a coin</h2>
            {currCoin}
            <button onClick={handleClick}>Flip!</button>
            <p>
                {headCount + tailCount} {headCount + tailCount === 1 ?  "flip" : "flips"} 
            </p>
            <p>
                 {headCount} {headCount === 1 ? "head" : "heads"} & {tailCount} {tailCount === 1 ? "tail" : "tails"} 
            </p>
        </div>
    );
}

CoinContainer.defaultProps = {
    coins: [
      {
        side: "head",
        imgSrc: "https://tinyurl.com/react-coin-heads-jpg"
      },
      {
        side: "tail",
        imgSrc: "https://raw.githubusercontent.com/Jayant818/Coin-Flipper/main/src/Tail.jpg"
      }
    ]
};

export default CoinContainer;