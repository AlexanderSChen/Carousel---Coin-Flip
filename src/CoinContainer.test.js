import React from 'react';
import { render, fireEvent, queryByAltText } from "@testing-library/react";
import CoinContainer from './CoinContainer';

beforeEach(function() {
    jest 
        .spyOn(Math, "random")
        .mockReturnValueOnce(0.25)
        .mockReturnValueOnce(0.75);
});

test("Smoke test", function() {
    render(<CoinContainer />);
});

test("Matches snapshot", function() {
    const { asFragment } = render(<CoinContainer />);
    expect(asFragment()).toMatchSnapshot();
});

test("Doesn't show coin when the page loads", function() {
    const { queryByTestId } = render(<CoinContainer />);

    expect(queryByTestId("coin")).toBeNull();
});

test("Counts correctly when coin flips heads", function() {
    const { getByText, queryByAltText } = render(<CoinContainer />);

    const button = getByText("Flip!");
    fireEvent.click(button);

    expect(queryByAltText("head")).toBeInTheDocument();
    expect(queryByAltText("tail")).not.toBeInTheDocument();
    expect(
        getByText("1 head & 0 tails")
    ).toBeInTheDocument();
});

test("Counts correctly when coin flips tails", function() {
    const { getByText, queryByAltText } = render(<CoinContainer />);

    const button = getByText("Flip!");
    fireEvent.click(button);
    fireEvent.click(button);

    expect(queryByAltText("tail")).toBeInTheDocument();
    expect(queryByAltText("head")).not.toBeInTheDocument();
    expect(
        getByText("1 head & 1 tail")
    ).toBeInTheDocument();
});

afterEach(function() {
    Math.random.mockRestore();
});