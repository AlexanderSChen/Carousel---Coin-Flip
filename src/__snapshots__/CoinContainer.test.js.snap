// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="CoinContainer"
  >
    <h2>
      Flip a coin
    </h2>
    <button>
      Flip!
    </button>
    <p>
      0 flips
    </p>
    <p>
      0 heads & 0 tails
    </p>
  </div>
</DocumentFragment>
`;
